### DISCUSSION TITLE
Trying to clone/pull private addons repositories using https

### DISCUSSION THREAD

**Discussion by gabonog** [Tecnativa/doodba#614] [Doodba - The image]
Hi everyone!
I am trying to implement a way to clone/pull private addons repositories via https securely. SSH is not an option because of technical limitations of the git server.

I tried something like this (and it worked) on `repos.yaml`:

```
./repo:
  defaults:
    depth: $DEPTH_DEFAULT
  remotes:
    origin: https://user:pass@git-server.tld/project/repo.git
  target: origin $ODOO_VERSION
  merges:
    - origin $ODOO_VERSION
```

But the problem here is to leave those credentials in plain text in the file. So, what I want to use is something similar to this (as suggested on point 3 here: https://github.com/Tecnativa/doodba/issues/287#issuecomment-583302443):

```
https://${USER}:${PASS}@git-server.tld/project/repo.git
```

But no idea how to implement that env vars and use them in `repos.yaml`. These are my failed attempts:

- In `common.yaml`: Add env vars in `build/args`.
- In `common.yaml`: Add env vars in `environment`.

Any suggestion to implement those env vars?
Thank you!

**Comment by ap-wtioit** [comment]
If your are talking about a development, i think the aggregate part is always called with setup-devel.yaml that does not extend common.yaml and therefore you will need to define `USER` and `PASS` in environment of setup-devel.yaml.

For prod.yaml and test.yaml setting them in common.yaml environment should work i think.

Also keep in mind [git-aggregator](https://github.com/acsone/git-aggregator/blob/master/git_aggregator/config.py#L172) is using [Template.substitute](https://docs.python.org/3/library/string.html), so i don't think `${USER}` will work but `$USER` should.

**Reply by ap-wtioit** [reply]
Ah and i forgot, avoid using `USER` it has a special meaning in linux and you might get unwanted side effects: https://unix.stackexchange.com/a/76356/222241

**Reply by pedrobaeza** [reply]
You can also use a deploy token instead.

**Reply by gabonog** [reply]
Thank very much you for your replies!
@ap-wtioit 
I want to use it in both test and prod, even in a CI in order to build the docker image. Oh... without square bracket... I will try ASAP and I leave a better explanation

@pedrobaeza 
That is exactly what I want to use but without "hard-coding" the token in the file. Maybe I am wrong and it is not that bad, but if it's possible, I prefer to use an env var that I can call when running the building command.

Thank you again for your answers!

**Reply by pedrobaeza** [reply]
Deploy tokens are for putting them in clear, and revoke as soon as they are compromised. They can also be limited for just read purposes.

**Reply by gabonog** [reply]
Thank you @ap-wtioit for the syntax warning when using variables in git-aggregator. That was very helpful.
Thank you @pedrobaeza for clarifying the security implications of putting tokens in code. The thing is also that we don't want to commit a change every time we want to modify a token.

I tried several things:

- Development
  - [x] Adding var to `odoo/environment` in `setup-devel.yaml` worked
- Prod/Test
  - [ ] Just adding the var as a `build-arg` to the build command expecting that will be available in the git-aggregator layer didn't work.
  - [ ] Adding some changes to the basic Dockerfile didn't work.

All these make sense because in `setup-devel.yaml` there is a `build-arg` `AGGREGATE: "false"` that avoids to run the git-aggregator when building but the `entrypoint` triggers it.

I am not able to reproduce this behaviour in prod/test buildings and I supposed it is not desired too. 
At this point, I think that the key is to make available the env var in the git-aggregator layer but I don't know how to do it.

Also, is it mandatory to remake the Dockerfile entirely? I am working from a copier template, so I can't just copy-paste the Dockerfile in Doodba repo... Can anybody guide me in how to customize the Dockerfile?

**Reply by ap-wtioit** [reply]
To enable custom credentials in Dockerfile you could use:
```Dockerfile
RUN --mount=type=secret,id=secret_id,target=/root/.netrc (umask 0027 && git clone --depth=$DEPTH_MERGE --branch=$ODOO_VERSION https://git-server.tld/path/to/odoo.git /opt/odoo/custom/src/odoo && cd /opt/odoo/custom/src/odoo && git remote remove origin)
```

with .netrc:
```.netrc
machine git-server.tld
login YOUR_USERNAME
password YOUR_TOKEN
```

and build args:
```
docker build ..... --secret id=secret_id,src=./.netrc
```

This enables .netrc credentials for a git clone (for root for odoo use /home/odoo/.netrc but that might have issues), and clone it into the expected directory and remove the remote in case the url is a secret as well.

Do not add .netrc files without the mount type secret as they might be contained in the docker layers then.

Edit: this applies if you build your own doodba images, not to a scaffolding project (one copied with copier)

**Reply by ap-wtioit** [reply]
To do build stuff in scaffolding projects place your scripts in odoo/custom/build.d, name them 099-my_script_adding_credentials to run before aggregate, 101-my_script_removing_credentials runs after aggregate. Keep mind to check environment variable `AGGREGATE` to have the correct behaviour for devel vs prod. Build stuff runs with root so give the created files correct permissions for odoo.

One more enhanced script that fiddles with repos.yaml and odoo users git settings can be found here: https://github.com/Tecnativa/doodba/blob/163ecb0f80b04605ed3b601bba441c7612f8564c/tests/scaffoldings/repo_merge/custom/build.d/099-git_merge_no_ff

All build scripts run in one RUN command so only one layer with the final result should be created, if you are careful that should not expose any secrets.

**Reply by gabonog** [reply]
Sounds good this last approach, but I can't make it work. So, to sum up:

File: `odoo/custom/src/repos.yaml`
```
# [...]

./repo:
  defaults:
    depth: $DEPTH_DEFAULT
  remotes:
    origin: https://$GIT_AUTH@git-server.tld/project/repo.git
  target: origin $ODOO_VERSION
  merges:
    - origin $ODOO_VERSION

# [...]
```

File: `odoo/custom/build.d/099-repos-add-auth`

```
#!/bin/bash
set -e

HARDCODED_AUTH="token"

sed -i 's/\$GIT_AUTH/'"$HARDCODED_AUTH"'/g' /opt/odoo/custom/src/repos.yaml
```
I am trying the simplest option but, when building, it is like the file doesn't exist, showing `KeyError: 'GIT_AUTH'` while running  `opt/odoo/common/build.d/100-repos-aggregate`.
Am I missing something?

Once this work, somehow I have to obtain the value from a `--build-arg` given on docker build command.

Again, thank you for the guidance.

**Reply by ap-wtioit** [reply]
does your odoo/custom/build.d/099-repos-add-auth file have execute permissions? `chmod +x odoo/custom/build.d/099-repos-add-auth`

**Reply by gabonog** [reply]
So embarrasing... It was execute permissions. Thank you!

**Reply by gabonog** [reply]
Finally my workaround is (based on https://github.com/Tecnativa/doodba/discussions/614#discussioncomment-9892679):

- Create those 2 files but changing `HARDCODED_AUTH="token"` to `HARDCODED_AUTH=$PIPELINE_AUTH` in `odoo/custom/build.d/099-repos-add-auth`.
- In the pipeline, before building the image, I run a `sed` to replace `PIPELINE_AUTH` in order have the value when running the script.

Thank you @ap-wtioit for helping me so much.
